### ISSUE TITLE
SIGILL in rust

### ISSUE THREAD

**Issue by qoega** [ClickHouse/ClickHouse#52053]
https://s3.amazonaws.com/clickhouse-test-reports/0/2ab41ed66559554dc396ef581e2f76511a0bbb6e/fuzzer_astfuzzertsan/report.html

```
(lldb) target create "clickhouse" --core "core"
Core file '/home/ubuntu/tmp/core' (x86_64) was loaded.
(lldb) bt
* thread #1, name = 'clickhouse-serv', stop reason = signal SIGILL
  * frame #0: 0x00005654d9022df6 clickhouse`core::ptr::drop_in_place$LT$std..backtrace_rs..symbolize..gimli..Mapping$GT$::h90d79bc21f4118dd at mod.rs:490:1
  * 
```

**Comment by alexey-milovidov** [comment]
@azat, where do you think it originated?

PS. I've heard that Rust does not have exception handling, and many libraries crash in normal scenarios (like not enough memory). If it is true, we cannot use Rust in the server.

**Comment by azat** [comment]
This backtrace looks wrong, @qoega how did you get it?

This is what I got:

```
wget https://s3.amazonaws.com/clickhouse-builds/23.7/2ab41ed66559554dc396ef581e2f76511a0bbb6e/package_tsan/clickhouse
wget https://s3.amazonaws.com/clickhouse-test-reports/0/2ab41ed66559554dc396ef581e2f76511a0bbb6e/fuzzer_astfuzzertsan/core.zst
chmod +x clickhouse
lldb clickhouse --core core*
(lldb) bt
(lldb) bt
* thread #1, name = 'clickhouse-serv', stop reason = signal SIGILL
  * frame #0: 0x00005654d9022df6 clickhouse`__sanitizer::CheckFailed(char const*, int, char const*, unsigned long long, unsigned long long) + 134
    frame #1: 0x00005654d90283e0 clickhouse`__sanitizer::StackStore::Alloc(unsigned long, unsigned long*, unsigned long*) + 256
    frame #2: 0x00005654d9028254 clickhouse`__sanitizer::StackStore::Store(__sanitizer::StackTrace const&, unsigned long*) + 100
    frame #3: 0x00005654d902a34b clickhouse`__sanitizer::StackDepotNode::store(unsigned int, __sanitizer::StackTrace const&, unsigned long long) + 43
    frame #4: 0x00005654d902a7d1 clickhouse`__sanitizer::StackDepotBase<__sanitizer::StackDepotNode, 1, 20>::Put(__sanitizer::StackTrace, bool*) + 561
    frame #5: 0x00005654d90aa26c clickhouse`__tsan::CurrentStackId(__tsan::ThreadState*, unsigned long) + 60
    frame #6: 0x00005654d9012be3 clickhouse`__sanitizer::DD::MutexInit(__sanitizer::DDCallback*, __sanitizer::DDMutex*) + 19
    frame #7: 0x00005654d90b57bd clickhouse`__tsan::DDMutexInit(__tsan::ThreadState*, unsigned long, __tsan::SyncVar*) + 93
    frame #8: 0x00005654d90c0913 clickhouse`__tsan::MetaMap::GetSync(__tsan::ThreadState*, unsigned long, unsigned long, bool, bool) + 323
    frame #9: 0x00005654d90b8bc9 clickhouse`__tsan::Release(__tsan::ThreadState*, unsigned long, unsigned long) + 105
    frame #10: 0x00005654d903865d clickhouse`__tsan::FdRelease(__tsan::ThreadState*, unsigned long, int) + 77
    frame #11: 0x00005654d904aca4 clickhouse`__interceptor_write + 244
    frame #12: 0x00005654e14e5bcd clickhouse`DB::WriteBufferFromFileDescriptorDiscardOnFailure::nextImpl(this=<unavailable>) at WriteBufferFromFileDescriptorDiscardOnFailure.cpp:16:23
    frame #13: 0x00005654d90f56ff clickhouse`DB::WriteBuffer::write(char const*, unsigned long) + 319
    frame #14: 0x00005654e177bc8b clickhouse`signalHandler(int, siginfo_t*, void*) [inlined] void DB::writePODBinary<DB::ThreadStatus*>(x=<unavailable>, buf=0x00007f7e77a32c10) at WriteHelpers.h:88:9
    frame #15: 0x00005654e177bc81 clickhouse`signalHandler(sig=<unavailable>, info=<unavailable>, context=<unavailable>) at BaseDaemon.cpp:159:5
    frame #16: 0x00005654d9045000 clickhouse`__tsan::CallUserSignalHandler(__tsan::ThreadState*, bool, bool, int, __sanitizer::__sanitizer_siginfo*, void*) + 256
    frame #17: 0x00005654d9045526 clickhouse`sighandler(int, __sanitizer::__sanitizer_siginfo*, void*) + 438
    frame #18: 0x00007f7feb44c520 libc.so.6`__ldexpf(value=<unavailable>, exp=<unavailable>) at s_ldexp_template.c:27:1

```

And it looks correct, since there are other errors from TSan in the log:

```
ThreadSanitizer: CHECK failed: sanitizer_stack_store.cpp:80 "((block_idx)) < (((sizeof(blocks_)/sizeof((blocks_)[0]))))" (0x1000, 0x1000) (tid=186)
```

But, back the the Rust.

I also concerned about integrating the Rust code into the server, since it is very hard to do this right, and using any crate like it did right now you are definitely cannot guarantee that some code does not `panic!` 

Rust does has unwinding, but it is tricky - https://doc.rust-lang.org/nomicon/unwinding.html

And about `panic!`, yes, there is a `![panic_handler]`, and for instance linux kernel overrides it - https://github.com/torvalds/linux/blob/b6e6cc1f78c772e952495b7416c9ac9029f9390c/rust/kernel/lib.rs#L92

But simply ignoring `panic` or convert it to C++ exception is not always OK, since there should be some real uncorrectable problem.

Linux kernel did a lot of job to make Rust safe for kernel, and obviously in the kernel environment you cannot use any random crate.

So to summarize, I think that it is not 100% safe to integrate Rust into the server, in the way it done right now at least. Since even simple allocation failures will lead to `panic!`, I guess.

**So in my opinion at least there should be a configuration option to enable Rust (and it should be set to OFF by default).**

_But I don't have enough expertise in Rust, so you can just ignore everything that I wrote._

**Comment by alexey-milovidov** [comment]
@azat The BLAKE3 library was added only to demonstrate the usage of Rust for the build system.
Currently, we also have Skim, and the BLAKE3 can be replaced with C or C++ version.

**Comment by azat** [comment]
Right now - yes. But there is also #50686 pending
But even one thing may crash the server actually, and it is better to hint this for users at least.

**Comment by antaljanosbenjamin** [comment]
Should the PRQL integration be put on hold until we figure this out?

**Comment by alexey-milovidov** [comment]
PRQL integration is ok.

If we find that it can lead to crashes, we will add it to the description of the setting.
In the worst-case scenario, we will make the setting read-only in ClickHouse Cloud.

**Comment by max-sixty** [comment]
Hi from a PRQL maintainer! Can I clarify — will panicking in rust cause the whole ClickHouse server to crash?

PRQL is much better than it used to be at not panicking. But even if we've closed the outstanding cases of it panicking, I suspect they're still out there.

So to the extent it causes the ClickHouse server to crash, we should prioritize robustness against panics on our end; e.g. some light fuzz testing.
